### PR TITLE
Press escape to desactivate drawing in edition

### PR DESCRIPTION
**Pull request by llienher**
https://jira.camptocamp.com/browse/GSEPF-259

### CODE REVIEW

File: src/editing/createfeatureComponent.js [hunk 1]
```diff
@@ -306,6 +314,9 @@ exports.Controller_.prototype.$onDestroy = function() {
     this.ngeoEventHelper_.clearListenerKey(uid);
     this.interaction_.setActive(false);
     this.map.removeInteraction(this.interaction_);
+    if (this.keyDownListener_) {
```
Comment (fredj) [inline src/editing/createfeatureComponent.js]: can be removed (see comment above)

File: src/editing/createfeatureComponent.js [hunk 2]
```diff
@@ -239,6 +239,14 @@ exports.Controller_.prototype.$onInit = function() {
   this.interaction_ = interaction;
   this.map.addInteraction(interaction);
 
+  this.keyDownListener_ = document.body.addEventListener('keydown', (evt) => {
```
Comment (fredj) [inline src/editing/createfeatureComponent.js]: It won't work: `addEventListener` doesn't return anything.

Use something like this instead:
```js
 this.ngeoEventHelper_.addListenerKey(
      uid,
      olEvents.listen(
        document.body,
        'keydown',
        this.handleKeyDown_,
        this
      )
    );
```

File: src/editing/createfeatureComponent.js [hunk 3]
```diff
@@ -239,7 +239,26 @@ exports.Controller_.prototype.$onInit = function() {
   this.interaction_ = interaction;
   this.map.addInteraction(interaction);
 
+  document.body.addEventListener('keydown', (evt) => {
+    const escPressed = evt.keyCode === 27; // Escape key
+    if (interaction.getActive() && escPressed) {
```
Comment (gberaudo) [inline src/editing/createfeatureComponent.js]:
```suggestion
    if (escPressed && interaction.getActive()) {
```

File: src/editing/createfeatureComponent.js [hunk 4]
```diff
@@ -264,6 +283,18 @@ exports.Controller_.prototype.$onInit = function() {
...
+ * Called when escape key is pressed to reset drawing.
+ * @param {ol.interaction.Draw.Event|ngeox.MeasureEvent} event Event.
+ * @export
+ */
+exports.Controller_.prototype.handleKeyDown_ = function(event) {
+  const escPressed = event.keyCode === 27; // Escape key
+  if (this.interaction_.getActive() && escPressed) {
+    this.interaction_.setActive(false);
+    this.interaction_.setActive(true);
+  }
```
Comment (gberaudo) [inline src/editing/createfeatureComponent.js]: It is surprising to have 2 times the same code in the same file.

File: contribs/gmf/src/drawing/drawFeatureComponent.js [hunk 5]
```diff
@@ -441,6 +441,8 @@ exports.Controller_.prototype.handleActiveChange_ = function(active) {
     }
   }
 
+  olEvents.unlistenByKey(this.cancelEventKey_);
```
Comment (gberaudo) [inline contribs/gmf/src/drawing/drawFeatureComponent.js]: Move in the else.

File: contribs/gmf/src/drawing/drawFeatureComponent.js [hunk 6]
```diff
@@ -592,6 +594,14 @@ exports.Controller_.prototype.handleMapClick_ = function(evt) {
     return;
   }
 
+  this.cancelEventKey_ = olEvents.listen(document.body, 'keydown', (e) => {
```
Comment (gberaudo) [inline contribs/gmf/src/drawing/drawFeatureComponent.js]: To be sure, check the this.cancelEventKey_ is not defined.

File: contribs/gmf/src/editing/editFeatureComponent.js [hunk 7]
```diff
@@ -987,6 +987,17 @@ exports.Controller_.prototype.handleMapSelectActiveChange_ = function(
  * @private
  */
 exports.Controller_.prototype.handleMapClick_ = function(evt) {
+  const interactions = evt.target.getInteractions().getArray();
+  const activeInteraction = interactions.find(interaction => interaction.getActive() === true);
+
+  this.cancelEventKey_ = olEvents.listen(document.body, 'keydown', (e) => {
```
Comment (gberaudo) [inline contribs/gmf/src/editing/editFeatureComponent.js]: idem

File: src/draw/Controller.js [hunk 8]
```diff
@@ -12,12 +12,16 @@ import ngeoMiscBtnComponent from 'ngeo/misc/btnComponent.js';
 import ngeoMiscDecorate from 'ngeo/misc/decorate.js';
 import ngeoMiscFeatureHelper from 'ngeo/misc/FeatureHelper.js';
 import olFeature from 'ol/Feature.js';
+import * as olBase from 'ol/index.js';
+import * as olEvents from 'ol/events.js';
```
Comment (gberaudo) [inline src/draw/Controller.js]: would be better to import only what you need.

File: src/draw/Controller.js [hunk 9]
```diff
@@ -139,7 +144,21 @@ const exports = function($scope, $sce, gettextCatalog,
...
   );
+};
+
 
+/**
+ * Called when escape key is pressed to reset drawing.
+ * @param {ol.interaction.Draw.Event|ngeox.MeasureEvent} event Event.
+ * @export
+ */
+exports.prototype.handleEscapeKeyDown_ = function(event) {
```
Comment (gberaudo) [inline src/draw/Controller.js]: Is it called?

File: src/editing/createfeatureComponent.js [hunk 10]
```diff
@@ -264,6 +275,18 @@ exports.Controller_.prototype.$onInit = function() {
...
 
+/**
+ * Called when escape key is pressed to reset drawing.
+ * @param {ol.interaction.Draw.Event|ngeox.MeasureEvent} event Event.
+ * @export
+ */
+exports.Controller_.prototype.handleEscapeKeyDown_ = function(event) {
+  const escPressed = event.keyCode === 27; // Escape key
+  if (escPressed && this.interaction_.getActive()) {
+    this.interaction_.setActive(false);
```
Comment (gberaudo) [inline src/editing/createfeatureComponent.js]: I think you should use a local variable.

File: src/draw/Controller.js (hunk 9 above)
Comment (llienher) [reply src/draw/Controller.js]: Line 197, in the listener

**Comment by sbrunner** [comment]
Is it relay without risque to merge it on a stable branch?

**Comment by gberaudo** [comment]
> Is it relay without risk to merge it on a stable branch?

@sbrunner, this is a new functionality that triggers only when the "esc" key is pressed.
Since that key does nothing at the moment the users will not notice the change.
So risk is low IMO.

**Comment by sbrunner** [comment]
Should be merged on 2.4 and master :-)

**Pull request merged**